### PR TITLE
Tutorial polar coordinates

### LINKED ISSUE #160: Document usage of kinematics in polar coordinates
PR #155 introduced kinematic variables in polar coordinates, and transformations between cartesian and polar coordinates. These features are being added to the API docs, but their usage is not yet documented in any example/tutorial.

Perhaps the best way to document this would be to amend the existing example about [computing and visualising kinematics](https://movement.neuroinformatics.dev/examples/compute_kinematics.html#sphx-glr-examples-compute-kinematics-py).

Alternatively, we could also split up the existing kinematics example into smaller bits (which ties into #70).

### PR DESCRIPTION
**Pull request by sfmig**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
We wanted to add a tutorial with an example use of polar coordinates.

**What does this PR do?**
\

## References
Closes #160 

## How has this PR been tested?
The docs build locally and on CI

## Is this a breaking change?

No

## Does this PR require an update to the documentation?

No

## Checklist:

- [x] The code has been tested locally
- [ n/a ] Tests have been added to cover all new functionality
- [ n/a ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: examples/compute_polar_coordinates.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+ax.set_ylabel("y (pixels)")
+ax.invert_yaxis()
+ax.set_title(f"Head trajectory ({mouse_name})")
+fig.colorbar(sc, ax=ax, label=f"time ({ds.attrs['time_unit']})")
+fig.show()
+
+# %%
+# We can see that the majority of the head trajectory data is within a
+# cruciform shape. This is because the dataset is of a mouse moving on an
+# [Elevated Plus Maze](https://en.wikipedia.org/wiki/Elevated_plus_maze).
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:97]: This link doesn't render as you'd expect, because the markup languages for these text blocks is rst not md. Use the appropriate sphinx rst syntax for hyperlinks, see https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

File: examples/compute_polar_coordinates.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+# compute the midpoint between the ears
+midpoint_ears = 0.5 * (
+    position.sel(keypoints="left_ear") + position.sel(keypoints="right_ear")
+)
+
+# compute the head vector
+head_vector = position.sel(keypoints="snout") - midpoint_ears
+
+# drop the keypoints dimension
+head_vector = head_vector.drop_vars("keypoints")
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:63]: it's not immediately clear why this is necessary, maybe a short explanation is warranted?

File: examples/compute_polar_coordinates.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+# %%
+# Notice how the resulting array has a ``space_pol`` dimension with two
+# coordinates: ``rho`` and ``phi``. These are the polar coordinates of the
+# head vector.
+#
+# The coordinate ``rho`` is the norm (length) of the vector. In our case, the
+# distance from the midpoint between the ears to the snout. The coordinate
+# ``phi`` is the orientation of the head vector relative to the positive
+# x-axis, and ranges from -``pi`` to ``pi``
+# (following the [atan2](https://en.wikipedia.org/wiki/Atan2) convention).
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:203]: same point about rendering hyperlinks in rst.

File: examples/compute_polar_coordinates.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+head_vector_polar = cart2pol(head_vector)
+
+print(head_vector_polar)
+
+# %%
+# Notice how the resulting array has a ``space_pol`` dimension with two
+# coordinates: ``rho`` and ``phi``. These are the polar coordinates of the
+# head vector.
+#
+# The coordinate ``rho`` is the norm (length) of the vector. In our case, the
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:199]: maybe also mention the term "magnitude". It's all the same, but some people maybe more familiar with one of these terms.

File: examples/compute_polar_coordinates.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    loc="best",
+)
+ax.set_ylabel("count")
+ax.set_xlabel(r"$\rho$" + " (pixels)")
+fig.show()
+
+# %%
+# We can see that there is some spread in the value of ``rho`` in this
+# dataset. This may be due to noise in the detection of the head keypoints,
+# or due to the mouse tipping its snout upwards during the recording.
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:257]: or downwards, right? (which is the more common occurrence in this particular video)

File: examples/compute_polar_coordinates.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+# compute number of bins
+bin_width_deg = 5  # width of the bins in degrees
+n_bins = int(360 / bin_width_deg)
+
+# remove NaN values from the phi data
+bool_not_nan = (
+    ~np.isnan(
+        head_vector_polar.sel(individuals=mouse_name, space_pol="phi").values
+    ),
+)
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:274]: would xarray's builtin [dropna](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.dropna.html) method also do the trick. Not very important, but we should take opprotunities to get people familiar with the xarray syntax whenever possible. Plus, it may end up being shorter

File: examples/compute_polar_coordinates.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+# compute histogram
+counts, bins = np.histogram(
+    head_vector_polar.sel(individuals=mouse_name, space_pol="phi").values[
+        bool_not_nan
+    ],
+    bins=np.linspace(-np.pi, np.pi, n_bins + 1),
+)
+
+
+# plot histogram as a bar plot in polar projection
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:285]: very cool plot btw!

Comment (niksirbi) [reply examples/compute_polar_coordinates.py:285]: I think this should be the thumbnail image for this example. Sphinx-gallery by default takes the first plot, but you can [configure that behaviour by adding a special comment](https://sphinx-gallery.github.io/stable/configuration.html#choosing-the-thumbnail-image).

File: examples/compute_polar_coordinates.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+ax.set_xticklabels([str(t) + "\N{DEGREE SIGN}" for t in xticks_in_deg])
+
+fig.show()
+
+# %%
+# The polar plot shows how in this small time window of 20 frames,
+# the head of the mouse moved from right to left.
+
+# %%
+# ``movement`` also provides a ``pol2cart`` convenience function to transform
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:358]: This bit should be under a separate subheading (since it doesn't belong to the "polar plot"). Alternatively you can also mention this earlier (e.g. when you first do the cart2po), up to you.

File: examples/compute_polar_coordinates.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+# %%
+# The ``phi`` circular histogram shows that the head vector appears at a
+# variety of orientations in this dataset.
+
+# %%
+# Polar plot of the head vector within a time window
+# ---------------------------------------------------
+# We can also use a polar plot to represent the head vector in time,
+# in a coordinate system centred at the head of the mouse. Again,
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:314]: I was a bit thrown off by this phrasing. We are are still in coordinates relative to the frame (allocentric) here, right? As in, the phi angles ar relative to the frame's x-axis vector.

Comment (niksirbi) [reply examples/compute_polar_coordinates.py:314]: Relatedly, when you say "from right to left" you mean "of the frame", right? Not necessarily "from the animal's right to the animal's left".

File: examples/compute_polar_coordinates.py (hunk 2 above)
Comment (niksirbi) [reply examples/compute_polar_coordinates.py:63]: Actually, I think there is a better, more "xarray"ic way to computed the midpoint, which will automatically get rid of the "keypoints" dim for you:
```python
midpoint_ears = position.sel(keypoints=["left_ear", "right_ear"]).mean(dim="keypoints")
```
Here, because we are explicitly taking the mean across the keypoints, xarray smartly collapses that dimension.

Comment (niksirbi) [reply examples/compute_polar_coordinates.py:63]: Oh but I realised, once you subtract the midpoint from the snout, the "keypoints" dim sneaks back in.
To avoid that, you could do:
```python
head_vector = position.sel(keypoints="snout").values - midpoint_ears
```
Though, this is necessarily more transparent than your version, so feel free to choose whichever you like best.

File: examples/compute_polar_coordinates.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+)
+
+# compute the head vector
+head_vector = position.sel(keypoints="snout") - midpoint_ears
+
+# drop the keypoints dimension
+head_vector = head_vector.drop_vars("keypoints")
+
+# %%
+# Visualise the head trajectory
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:66]: I wonder if this is a good opportunity to showcase the ability to plot the frame under the trajectory (now that we have the frame for this dataset).
You could first load and plot the image with maptlotlib, then overplot the trajectory over it. It will also be a good visual check that the coordinates of the trajectory match the image, and the reason for the cruciform shape will be obvious.

You can get the frame path in a similar way to the poses path:
```python
frame_path = sample_data.fetch_dataset_paths("SLEAP_single-mouse_EPM.analysis.h5")["frame"]
```
If you had loaded the sample data the "normal" way, i.e. via `fetch_dataset`, the frame path would already be among the dataset attributes.

File: examples/compute_polar_coordinates.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+# area of interest
+xmin, ymin = 600, 665  # pixels
+x_delta, y_delta = 125, 100  # pixels
+
+# time window
+time_window = range(1650, 1670)  # frames
+
+
+# %%
+# For that subset of the data, we now plot the head vector.
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:115]: The code block that follows is quite long. If you want (completely optional) sphinx-gallery allows you to [split a figure's code across multiple blocks](https://sphinx-gallery.github.io/dev/configuration.html#defer-figures).

File: examples/compute_polar_coordinates.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+)
+
+ax.axis("equal")
+ax.set_xlim(xmin, xmin + x_delta)
+ax.set_ylim(ymin, ymin + y_delta)
+ax.set_xlabel("x (pixels)")
+ax.set_ylabel("y (pixels)")
+ax.set_title(f"Zoomed in to check ({mouse_name})")
+ax.invert_yaxis()
+fig.colorbar(sc, ax=ax, label=f"time ({ds.attrs['time_unit']})")
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:166]: the colorbar contains decimal-values ticks (.0 and .5) for some reason. That's confusing because the unit is frames, so it should be all integers, right?

File: examples/compute_polar_coordinates.py [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+# ----------------------------
+# Since ``rho`` is the distance between the ears' midpoint and the snout,
+# we would expect ``rho`` to be approximately constant in this data. We can
+# check this by plotting a histogram of its values across the whole clip.
+
+fig, ax = plt.subplots(1, 1)
+
+rho_data = head_vector_polar.sel(individuals=mouse_name, space_pol="rho")
+
+# compute and plot histogram
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:221]: btw, xarray also has a built-int histogram function, in case you want to try it. You can see it being used in the filtering example.

**Review by niksirbi — APPROVED** [review]
Thanks a lot @sfmig :tada:!

This looks good, I just had a few clarifying questions and some optional suggestions.

Also, two thoughts came to mind:
- @b-peri is working on implementing head direction as a derived variable. Once we have that, we should edit the first half of this example to use our implementation.
- In your examples, you go at great lengths to produce detailed and informative plots, but this take lots of matplotlib code. Perhaps we should wrap the most useful such plots as `movement.plot` utilities, for our (and the users) convenience. Trajectory is the obvious one here. To be clear, this is feature idea for the future (not for this PR).

File: examples/compute_polar_coordinates.py (hunk 5 above)
Comment (sfmig) [reply examples/compute_polar_coordinates.py:257]: but downwards wouldn't detect the snout right?

Comment (niksirbi) [reply examples/compute_polar_coordinates.py:257]: yeap, in that case it would disappear.

File: examples/compute_polar_coordinates.py (hunk 2 above)
Comment (sfmig) [reply examples/compute_polar_coordinates.py:63]: thanks! I like that we show `xarray` methods so went for the following
- using the `mean()` approach
- using `drop_vars()`, but also added more context in the comment.

File: examples/compute_polar_coordinates.py (hunk 11 above)
Comment (sfmig) [reply examples/compute_polar_coordinates.py:148]: oh nice, thanks! I think in this case it may not be the best: the resulting jupyter notebook plots the intermediate plots, which would then require a few more lines of code to format axes etc for each of them. 

But will keep it in mind for the next tutorial. I was actually trying to do exactly this for another plot, in which I showed the selected spatial region and time window of interest in the head trajectory plot (but I ended up scrapping that plot altogether to reduce length).

File: examples/compute_polar_coordinates.py (hunk 9 above)
Comment (sfmig) [reply examples/compute_polar_coordinates.py:344]: Yeah, I think I tried to say it in simpler terms / less words and it was worse 😅 . 

In the following when I say "head centre" I mean "midpoint between the ears". I am also assuming the keypoints in the head of the mouse are a rigid body (that is, at constant distances).

> We are are still in coordinates relative to the frame (allocentric) here, right? As in, the phi angles ar relative to the frame's x-axis vector.

The head vector in this plot is represented in a coordinate system parallel to the pixel-space coordinate system but with it's origin at the mouse's head centre, at every frame. So as we go through frames, this "head coordinate system" is parallel to the pixel-space coord system, but its origin translates with the mouse's head centre. 

Since both coordinate systems are parallel, the theta angle computed relative to the x-axis in pixel-space is the same as the theta angle for the head coordinate system (because both x-axes are parallel). Does that make sense?

> Relatedly, when you say "from right to left" you mean "of the frame", right? Not necessarily "from the animal's right to the animal's left".

Well, in this case it would be the same, because both coordinate systems are always parallel. But I agree "left" and "right" of the animal/frame is ambiguous (depends on what is defined as forward right? I was trying to keep it simple sorry!).

I think the confusion comes from calling this "a coordinate system centred at the head of the mouse". While this is correct, it doesn't specify how this coordinate system is oriented at every frame. There are two coordinate systems that move with the mouse's head that are most used in this context:
- one would be a coordinate system whose origin is always at the mouse's head, but that is always parallel to the pixel-space coordinate system. Note that in this coordinate system, the snout of the mouse (or any other keypoint in the head) will change position as the mouse moves its head. This is what is represented in the plot (the head centre would be the origin, and the snout would be the markers).
- another coordinate system is one whose orientation is constant relative to the mouse's head. This coordinate system may be defined at every frame with its origin at the mouse's head centre, it's positive x-axis going from the head centre to the right ear, and it's y-axis perpendicular to the x-axis and positive going from the head centre to the snout. This coordinate system is "linked" to the head of the mouse at every frame: in this coordinate system the snout will be at a constant location throughout the clip (assuming the head is a rigid body). I think this is usually called 'egocentric' in neuroscience? I think you assumed this one right?

Will clarify in the text cell!

File: examples/compute_polar_coordinates.py [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+sc = ax.scatter(
+    phi,
+    np.ones_like(phi),  # assign a constant value rho=1 for visualization
+    c=time_window,
+    cmap="viridis",
+    s=50,
+)
+
+# axes settings
+ax.set_theta_direction(-1)  # set direction counterclockwise
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:366]: the comment here says "counterclockwise", but in the previous plot the same line is designated as "clockwise". One of them has to be wrong.

File: examples/compute_polar_coordinates.py (hunk 9 above)
Comment (niksirbi) [reply examples/compute_polar_coordinates.py]: Thanks for the clarification, I actually assumed the one you meant (the first option), it's just that the way it was phrased made me doubt it. I'm fine with the way it's worded now.
And correct, the 2nd coordinate system you refer to is the one referred to as "egocentric" (which is not what we have here).

File: examples/compute_polar_coordinates.py [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+    c=midpoint_ears.time,
+    cmap="viridis",
+    marker="o",
+    alpha=0.05,  # transparency
+)
+
+ax.axis("equal")
+ax.set_xlabel("x (pixels)")
+ax.set_ylabel("y (pixels)")
+ax.invert_yaxis()
```
Comment (niksirbi) [inline examples/compute_polar_coordinates.py:127]: I think there is no need for the y axis inversion this time. 
Because you plot the image first, matplotlib shows it in the "image" coordinate system (with y axis values increasing from top to bottom). When you overplot the trajectory on it, the "image" orientation is still maintained. Your axis inversion here undoes it and turns it the "wrong" way around. You can see that if you look at the y axis ticks of this vs the previous plot. If you verify this, maybe it's worth leaving a small comment to that effect (i.e. y axis inversion not needed if image is plotted first).

**Review by niksirbi — APPROVED** [review]
Thanks a lot @sfmig !!
I'm very happy with how this looks now, just two minor comments, and after that ready to merge.
